plugins/org.eclipse.n4js/src/org/eclipse/n4js/scoping/N4JSScopeProvider.xtend

```diff
@@ -142,40 +152,62 @@ class N4JSScopeProvider extends AbstractScopeProvider implements IDelegatingScop
 		return delegate;
 	}
 
+	/** Dispatches to concrete scopes based on the context and reference inspection */
 	override getScope(EObject context, EReference reference) {
 		try {
-			if (reference == TypeRefsPackage.Literals.PARAMETERIZED_TYPE_REF__DECLARED_TYPE) {
-				return new ValidatingScope(getTypeScope(context, reference, false),
-					context.getTypesFilterCriteria(reference));
+			// dispatch based on language variant
+			val resourceType = ResourceType.getResourceType(context)
+			switch (resourceType) {
+				case ResourceType.N4JSX	: return getN4JSXScope(context, reference)
+				case ResourceType.JSX	: return getN4JSXScope(context, reference)
+				default					: return getN4JSScope(context, reference)
 			}
 
-			// early catch:
-			if (reference.EReferenceType == N4JSPackage.Literals.LABELLED_STATEMENT) {
-				return scope_LabelledStatement(context);
-			}
-
-			// otherwise use context:
-			switch (context) {
-				ImportDeclaration:
-					return scope_ImportedModule(context, reference)
-				NamedImportSpecifier:
-					return scope_ImportedElement(context, reference)
-				IdentifierRef:
-					return scope_IdentifierRef_id(context, reference)
-				ParameterizedPropertyAccessExpression:
-					return scope_PropertyAccessExpression_property(context, reference)
-				N4FieldAccessor:
-					return Scopes.scopeFor(EcoreUtil2.getContainerOfType(context, N4ClassifierDefinition).ownedFields)
-			}
 		} catch (Error ex) {
-			if (context!==null && context.eResource.errors.empty) {
+			if (context !== null && context.eResource.errors.empty) {
 				throw ex;
 			}
-			// swallow exception, we got a parse error anyway
+		// swallow exception, we got a parse error anyway
+		}
+
+		return IScope.NULLSCOPE;
+	}
+
+	private def getN4JSScope(EObject context, EReference reference) {
+		// maybe can use scope shortcut
+		val maybeScopeShortcut = getScopeByShortcut(context, reference);
+		if (maybeScopeShortcut !== IScope.NULLSCOPE)
+			return maybeScopeShortcut;
+
+		// otherwise use context:
+		return getScopeForContext(context, reference)
+	}
+
+	/** shortcut to concrete scopes based on reference sniffing. May return {@link IScope#NULLSCOPE} if no suitable scope found */
```

Comment:
`Will return ... iff` instead of `May return ... if`.

Reply:
changed